wdl-engine/src/http/azure.rs

```diff
@@ -0,0 +1,194 @@
+    if url.scheme() != "https" {
+        return false;
+    }
+
+    if let Some(url::Host::Domain(domain)) = url.host() {
+        if let Some(account) = domain.strip_suffix(AZURE_STORAGE_DOMAIN_SUFFIX) {
+            // If the URL already has a query string, don't modify it
+            if url.query().is_some() {
+                return true;
+            }
+
+            if let Some(mut segments) = url.path_segments() {
+                // Determine the container name; if there's only one path segment, then use the
+                // `$root` container
+                let container = match (segments.next(), segments.next()) {
+                    (Some(_), None) => "$root",
+                    (Some(container), Some(_)) => container,
+                    _ => return true,
+                };
+
+                if let Some(containers) = config.auth.get(account) {
+                    if let Some(token) = containers.get(container) {
+                        let token = token.strip_prefix('?').unwrap_or(token);
+                        url.set_query(Some(token));
+                    }
+                }
+            }
+
+            return true;
+        }
+    }
+
+    false
+}
+
+#[cfg(test)]
+mod test {
+    use std::collections::HashMap;
+
+    use pretty_assertions::assert_eq;
+
+    use super::*;
+
+    #[test]
+    fn it_rewrites_urls() {
+        let url = rewrite_url(&"az://foo/bar/baz".parse().unwrap()).unwrap();
+        assert_eq!(url.as_str(), "https://foo.blob.core.windows.net/bar/baz");
+
+        let url = rewrite_url(&"az://foo/bar/baz#qux".parse().unwrap()).unwrap();
+        assert_eq!(
+            url.as_str(),
+            "https://foo.blob.core.windows.net/bar/baz#qux"
+        );
+
+        let url = rewrite_url(&"az://foo/bar/baz?qux=quux".parse().unwrap()).unwrap();
+        assert_eq!(
+            url.as_str(),
+            "https://foo.blob.core.windows.net/bar/baz?qux=quux"
+        );
+
+        let url =
+            rewrite_url(&"az://foo/bar/baz?qux=quux&jam=cakes#frag".parse().unwrap()).unwrap();
+        assert_eq!(
+            url.as_str(),
+            "https://foo.blob.core.windows.net/bar/baz?qux=quux&jam=cakes#frag"
+        );
+    }
+
+    #[test]
+    fn it_applies_auth() {
+        let mut config = AzureStorageConfig::default();
+        config.auth.insert(
+            "account".to_string(),
+            HashMap::from_iter([
+                ("container1".to_string(), "token1=foo".to_string()),
+                ("container2".to_string(), "?token2=bar".to_string()),
+                ("$root".to_string(), "token3=qux".to_string()),
+            ]),
+        );
+
+        // Not an Azure URL
+        let mut url = "https://example.com/bar/baz".parse().unwrap();
+        assert!(!apply_auth(&config, &mut url));
+        assert_eq!(url.as_str(), "https://example.com/bar/baz");
+
+        // Not using HTTPS
+        let mut url = "http://foo.blob.core.windows.net/bar/baz".parse().unwrap();
```

Comment:
Maybe I'm missing something, but are these truly failing to apply auth because of the reasons stated? E.g., if you add `https` here, will it then add the token? This URL doesn't look like the ones further below where it adds the token. Same with the next one I believe.

Reply:
(Later) if this needs to be updated, then obviously the same places in the Google/AWS parts need updating as well.